pybandits/mab.py

```diff
@@ -0,0 +1,398 @@
+from pybandits.model import Model
+from pybandits.strategy import Strategy
+from pybandits.utils import extract_argument_names_from_function
+
+
+class BaseMab(PyBanditsBaseModel, ABC):
+    """
+    Multi-armed bandit superclass.
+
+    Parameters
+    ----------
+    actions : Dict[ActionId, Model]
+        The list of possible actions, and their associated Model.
+    strategy : Strategy
+        The strategy used to select actions.
+    epsilon : Optional[Float01], 0 if not specified.
+        The probability of selecting a random action.
+    default_action : Optional[ActionId], None if not specified.
+        The default action to select with a probability of epsilon when using the epsilon-greedy approach.
+        If `default_action` is None, a random action from the action set will be selected with a probability of epsilon.
+    """
+
+    actions: Dict[ActionId, Model]
+    strategy: Strategy
+    epsilon: Optional[Float01] = None
+    default_action: Optional[ActionId] = None
+
+    def __init__(
+        self,
+        actions: Dict[ActionId, Model],
+        epsilon: Optional[Float01] = None,
+        default_action: Optional[ActionId] = None,
+        **strategy_kwargs,
+    ):
+        if "strategy" in strategy_kwargs:
+            strategy = strategy_kwargs["strategy"]
+            if len(strategy_kwargs) > 1:
+                raise ValueError("strategy should be the only keyword argument.")
+        else:
+            strategy_class = self.model_fields["strategy"].annotation
+            strategy = strategy_class(**strategy_kwargs)
+
+        super().__init__(actions=actions, strategy=strategy, epsilon=epsilon, default_action=default_action)
+
+    ############################################ Instance Input Validators #############################################
+
+    @field_validator("actions", mode="before")
+    @classmethod
+    def at_least_one_action_is_defined(cls, v):
+        # validate number of actions
+        if len(v) == 0:
+            raise AttributeError("At least one action should be defined.")
+        elif len(v) == 1:
+            warnings.warn("Only a single action was supplied. This MAB will be deterministic.")
+        # validate that all actions are of the same configuration
+        action_models = list(v.values())
+        first_action = action_models[0]
+        first_action_type = type(first_action)
+        if any(not isinstance(action, first_action_type) for action in action_models[1:]):
+            raise AttributeError("All actions should follow the same type.")
+
+        return v
+
+    @model_validator(mode="after")
+    def check_default_action(self):
+        if not self.epsilon and self.default_action:
+            raise AttributeError("A default action should only be defined when epsilon is defined.")
+        if self.default_action and self.default_action not in self.actions:
+            raise AttributeError("The default action should be defined in the actions.")
```

Comment:
Suggestion:
`raise AttributeError("The default action must be valid action defined in the actions set.")`